login: Add Skip TLS Verify checkbox

daniel-shuy opened the pull request:
Partially implements #50

Adds a `Skip TLS Verify` checkbox to the login dialog that will add [--insecure-skip-tls-verify](https://docs.openshift.com/online/pro/dev_guide/authentication.html) to `oc login` if enabled:
![image](https://user-images.githubusercontent.com/17351764/195652604-467a9fad-efb5-45e5-bfbc-42d8774ca0d9.png)

Because the [--insecure-skip-tls-verify](https://docs.openshift.com/online/pro/dev_guide/authentication.html) option needs to be added to all `oc` commands (thanks to @fbricon for highlighting this), the `boolean` state needs to be stored globally, not just in the login dialog. As such, I've used [recoiljs](https://recoiljs.org/) to do so, to be consistent with the existing implementation (e.g. `KubeContext`).

daniel-shuy commented on client/src/ContextCard.tsx:
```diff
@@ -137,6 +141,7 @@ export default function CurrentContext() {
               </Tooltip>
             )}
           </Box>
+          <Box><b>Skip TLS Verify:</b> {ocOptions.skipTlsVerify}</Box>
```
I modified the `ContextCard` to display the current `skipTlsVerify` state, is this alright?

fbricon commented:
It looks like the flag is only stored for the current session, can you use local storage to store a map of contextUrls/options pairs, so it can survive restarts? https://github.com/redhat-developer/openshift-dd-ext/blob/e9650640f2aeff8cceae375cee3081dd27ba0b6f/client/src/hooks/useStorageState.ts#L52

daniel-shuy commented:
@fbricon I've added the usage of local storage to persist the `OcOption`s as you suggested, but I mapped them by the context name instead of the cluster URL, since the context name is the unique identifier according to `OcUtils.loadContextUiData()`, 
https://github.com/redhat-developer/openshift-dd-ext/blob/220ebf0ae4a155a0a376153a720d024e8060a9e1/client/src/utils/OcUtils.ts#L147-L148

There might be multiple contexts with the same cluster URL, and the `OcOption`s might also have more options that are not cluster-specific in the future.

fbricon commented:
PR doesn't compile:

```
Failed to compile.

TS2322: Type 'void | undefined' is not assignable to type 'ExecProcess | undefined'.
  Type 'void' is not assignable to type 'ExecProcess | undefined'.
    386 |
    387 | function ocStream(ocOptions: OcOptions, args: string[], stream: ExecStreamOptions): ExecProcess | undefined {
  > 388 |   return ddClient.extension?.host?.cli.exec(ocPath, args, {
        |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  > 389 |     stream,
        | ^^^^^^^^^^^
  > 390 |   });
        | ^^^^^^
    391 | }
    392 |
    393 | function handleError(reject: (reason?: any) => void, e: any) {


error Command failed with exit code 1.
```

daniel-shuy commented:
@fbricon oops turns out it was because I ran `npm install` before `yarn install`, so it was using the wrong version of `@docker/extension-api-client-types` in my local environment (the `yarn.lock` was not used). I've fixed it and also rebased onto latest `main` branch.

fbricon commented:
Even though I checked the checkbox on login, the Skip TLS Verify option doesn't show on the context card:
<img width="948" alt="Screenshot 2022-10-24 at 12 05 02" src="https://user-images.githubusercontent.com/148698/197502311-250d158a-86be-4e68-98db-a55c2e6b6471.png">

I think it should be a checkbox on that card, so users can toggle it on/off, but it might not be something that people do in real life, so I'm not hung up on that.

daniel-shuy commented:
Oops I forgot that I need to convert the `boolean` to a `String` for React to render it, fixed!

daniel-shuy commented:
> I think it should be a checkbox on that card, so users can toggle it on/off, but it might not be something that people do in real life, so I'm not hung up on that.
Hmm I wanted to keep the current structure of the `ContextCard` being display-only, might seem a bit counterintuitive to allow changing the Skip TLS Verify but not the other login settings from the `ContextCard`

fbricon commented:
It's OK, I think we should provide a way to edit a given connection anyway, so this can be done in another PR

fbricon commented:
Thanks @daniel-shuy!
